streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.kafka.streams.kstream.Initializer;
+import org.apache.kafka.streams.kstream.KTable;
+import org.apache.kafka.streams.kstream.Merger;
+import org.apache.kafka.streams.kstream.SessionWindows;
+import org.apache.kafka.streams.kstream.Window;
+import org.apache.kafka.streams.kstream.Windows;
+import org.apache.kafka.streams.kstream.internals.graph.ProcessorParameters;
+import java.util.Collections;
+import org.apache.kafka.streams.kstream.internals.graph.StatefulProcessorNode;
+import org.apache.kafka.streams.kstream.internals.graph.StreamsGraphNode;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+
+class CogroupedStreamAggregateBuilder<K, VOut> {
+
+
+    private final InternalStreamsBuilder builder;
+
+    CogroupedStreamAggregateBuilder(final InternalStreamsBuilder builder) {
+        this.builder = builder;
+    }
+
+    <KR, VIn, W extends Window> KTable<KR, VOut> build(
+            final Map<KGroupedStreamImpl<K, ?>, Aggregator<? super K, ? super Object, VOut>> groupPatterns,
+            final Initializer<VOut> initializer,
+            final NamedInternal named,
+            final StoreBuilder<? extends StateStore> storeBuilder,
+            final Serde<KR> keySerde,
+            final Serde<VOut> valSerde,
+            final String queryableName,
+            final Windows<W> windows,
+            final SessionWindows sessionWindows,
+            final Merger<? super K, VOut> sessionMerger) {
+
+        final Collection<? extends AbstractStream<K, ?>> groupedStreams = new ArrayList<>(groupPatterns.keySet());
+        final AbstractStream<K, ?> kGrouped = groupedStreams.iterator().next();
+        groupedStreams.remove(kGrouped);
+        kGrouped.ensureCopartitionWith(groupedStreams);
+
+
+        final Collection<StreamsGraphNode> processors = new ArrayList<>();
+        boolean stateCreated = false;
+        int counter = 0;
+        for (final Entry<KGroupedStreamImpl<K, ?>, Aggregator<? super K, ? super Object, VOut>> kGroupedStream : groupPatterns
+                .entrySet()) {
+            final StatefulProcessorNode statefulProcessorNode = getStatefulProcessorNode(
+                    kGroupedStream.getValue(),
+                    initializer,
+                    named.withName(named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME) + "-" + counter++),
+                    stateCreated,
+                    storeBuilder,
+                    windows,
+                    sessionWindows,
+                    sessionMerger);
+            stateCreated = true;
+            processors.add(statefulProcessorNode);
+            builder.addGraphNode(kGroupedStream.getKey().streamsGraphNode, statefulProcessorNode);
+        }
+        final String functionName = named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME);
```

Comment:
nit: why `functionName` ? To be more explicit, maybe `mergeProcessorName` ?

Reply:
I reran the `useNamed()` test and the processors are now named `test-0` (for the aggregation processor) and `test` (for the merge processor) -- this will be hard for people to reason about. I would suggest to add more hard-coded additional information to get names like `test-cogroup-1` (or maybe `test-agg-1`?) and `test-merge` -- the names prefix that the user provides tires all those processors together and the user can identify that all those processor belong the `aggregate()` -- the additional suffix information helps to give more context what each processor does?

For `join` we do something similar (cf. `KStreamImplJoin#join()` method using `NamedInternal#suffixWithOrElseGet(...)`

Reply:
used suffixWithOrElseGet it fixed the problems I think